Section 6 testability statements & section 7 requirement numbers update (#210)
eea-oasis/baseline-standard: core/baseline-core-v1.0-psd01.md

 -->

This specification is related to:

**[baseline-api-v1.0]** _Baseline API and Data Model Version 0.1_. Edited by Dr. Andreas Freund, Anais Ofranc and Kyle Thomas. 17 November 2021. OASIS Standard. https://github.com/eea-oasis/baseline-standard/blob/main/api/baseline-api-v1.0-psd01.md . Latest stage: https://github.com/eea-oasis/baseline-standard/blob/main/api/baseline-api-v1.0-psd01.md.

**[baseline-ccsm-v1.0]** _Baseline CCSM Requirements Version 0.1_. Edited by Dr. Andreas Freund, Anais Ofranc and Kyle Thomas. 17 November 2021. OASIS Standard. https://github.com/eea-oasis/baseline-standard/blob/main/ccsm/baseline-ccsm-v1.0-psd01.md . Latest stage: https://github.com/eea-oasis/baseline-standard/blob/main/ccsm/baseline-ccsm-v1.0-psd01.md.

#### Abstract:
This document describes the minimal set of business and technical prerequisites, functional and non-functional requirements, together with a reference architecture that when implemented ensures that two or more Systems of Record can synchronize their system state over a Consensus Controlled State Machine (CCSM) network with little or no trust assumptions.

#### Status:

> This document is under active development and implementers are advised against implementing the specification unless they are directly involved with the Baseline Core Dev team.

Author: They are two core devs who are standard editors ... will update accordingly.

A Consensus Controlled State Machine (CCSM) is a network of replicated, shared, and synchronized digital data spread across multiple sites connected by a peer-to-peer and utilizing a consensus algorithm. There is no central administrator or centralized data storage.

**Data Orchestration:** 

An automated process for taking siloed data from multiple storage locations, combining and organizing it, and making it available for analysis.

> 

Author: happened by accident. reinstated.

An individual, organization, or company that has legal rights and obligations.

**Low latency:**

A latency that does not Materially Impact the overall system latency of the BPI.

> 

Author: happened by accident. reinstated

*This is just a well-known convention from business process management frameworks.*

[[R211]](#r211) Testability: A valid workstep is a function that transforms a given input, using one or more transformations of the input into an output. An invalid workstep either does not have an input, or transformations of the input into an output, or an output, or any combination of the three components.

#### **[R212]** 

> The input of a workstep MUST represent a new, proposed state of a state object compliant with the agreement between the agreement counterparties.

Reviewer: Seems like this is conflating at least two things: 
- The output must represent a new valid state object different from the input (the fact that it must be different repeats the transformation requirement in R211)
- The output must conform to the agreement (which is also what R213 says)

Author: I think the new wording you suggested is fine.

> [[R212]](#r212) Testability: The input of a workstep where the workstep is represented as a function as in the testability of [[R211]](#r211) can be formatted in a (data) schema of a state object that represents the state of a given agreement between counterparties such as the accumulated amount billed by a Seller to a Buyer. A function input formatted in this way can be validated by the function's logic to be a) conformant to the agreed upon state object schema, and b) to not be equal to the current state object. And the result of these two comparisons can be expressed as a true/false boolean statement as the function output. Therefore, the input can be evaluated to be valid -- a new state object (proposal) -- or not as per the requirement.  

Reviewer: Please simplify, but noting comment on the requirement itself which might be why this is complex. Partial suggestion depending on how to interpret the overall requirement:

```suggestion
[[R212]](#r212) Testability: The input and output of a workstep can be formatted according to the data schema of a state object that represents the state of a given agreement.

If the input and output are valid representations of a state according to the agreement, and are not equal, they meet the requirement.

If they are equal, or cannot be represented as a valid state object, they do not meet the requirement.
```

*See sections 6.4 and 6.5 below on BPI Accounts associated with state objects and BPI transactions for details.*

[[R214]](#r214) Testability: IF an input is provided to a workstep, THEN the output of the workstep MUST represent the validation result of the input as a correct or incorrect new agreement state. AND the validation result MUST be verifiable. To verify the validation result, the output of the workstep should contain information that allows the result to be tested or confirmed using a predefined method or standard. This may involve providing a reference to the criteria or rules used for validation, or specifying the expected behavior of the new agreement state. In addition, the requirement implies that the output of the workstep should only represent the validation result of the input, and not any other information or data. Therefore, a test should ensure that the output is limited to this specific purpose and does not include any extraneous or unnecessary information. 

#### **[R215]** 

> A workstep instance MUST be associated with only one workgroup.

Reviewer: Why? Given that they meet the other requirements, why not allow composability?

Author: Fair enough. Removed the constraining language at the end of the statement.

> [[R215]](#r215) Testability: A test of this requirment can be comprised of the following sequence:

Reviewer: ```suggestion
[[R215]](#r215) Testability: A test of this requirement can be comprised of the following sequence:
```

5. Attempt to create a work step instance without specifying a workgroup.
6. Verify that the system prompts the user to select a workgroup before proceeding with the creation of the work step instance.
7. Verify that the system prevents the creation of a work step instance without associating it with a workgroup by creating an error warning and not creating the work step instance.
8. Attempt to associate a work step instance with a non-existent workgroup.
9. Verify that the system prompts the user to select a valid workgroup before proceeding with the association of the work step instance by creating an error warning and not proceeding with the system process..
10. Verify that the system prevents the association of a work step instance with a non-existent workgroup by creating an error warning and aborting the system process.  

If all of the above test steps are successfully completed, then the requirement is met

#### **[R216]** 
A workstep instance MUST inherit the security and privacy policies of its associated workgroup.

> *See the details of workgroups and their security and privacy policies below.*

Reviewer: "Below" where? Link please.

Author: done.

#### **[R217]** 
A workstep MUST have a unique identifier within a BPI.

[[R217]](#r217) Testability:

Preconditions:

* A BPI test instance is set up and operational.

> * There are at least two worksteps set-up within the BPI.

Reviewer: If there is only one step, just check that there is an identifier.

Author: Not quite. It says unique identifier (within the BPI namespace). So you need at least two to check that identifiers are different and generating two with the same identifier is not permitted.

Test steps:

1. Retrieve the list of worksteps within the BPI.
2. For each workstep in the list, validate that the workstep identifiers are not duplicate.

> 3. Create a new workstep within the BPI and verify that it's identifier does not exist in the list of other workstep identifiers.

Reviewer: Why? The requirement only talks about worksteps that are in the BPI, not what happens if you make changes. And if you do, then you test again with the existing procedure, since otherwise you don't know if you conform any more.

It's possible to carry out that process by having a well-defined change process that incorporates a test or a mechanism to ensure that it will pass the test, but it is not necessary.

Reviewer: (also, "it's" -> "its", in the current text)

Author: This is just checking uniqueness ... for that you must create another workstep (all in the BPI of course)

> 4. Attempt to create a new workstep within the BPI with the same identifier as an existing workstep and verify that it is rejected.

Reviewer: see comment on step 3.

Author: same reply as for step 3.

Test Passing criteria:

* Test step 2 passes if all worksteps in the list have a unique identifier within the BPI.
* Test step 3 passes if the new workstep is assigned a unique identifier.
* Test step 4 passes if the attempt to create a workstep with the same identifier as an existing workstep is rejected.

#### **[R218]** 
A workstep MUST be updatable.

[[R218]](#r218) Testability:

Preconditions:

* A BPI test instance is set up and operational.
* A workstep exists in the BPI system.

Test Steps:

1. Access the workstep in the BPI system.
2. Attempt to update the workstep with new information.
3. Save the changes to the workstep.

> 4. AAccess the workstep in the BPI system again.

Reviewer: ```suggestion
4. Access the workstep in the BPI system again.
```

5. Verify that the new workstep information displays correctly

Test Passing Criteria:

* The changes to the workstep are saved successfully and can be viewed when accessing the workstep again.

#### **[R219]** 
A workstep instance MUST NOT be updated while the workstep instance is being executed by the BPI.

*This ensures that no breaking changes with potentially significant negative business impact are introduced while a workstep instance is being executed.*

[[R219]](#r219) Testability:

Preconditions:

* A BPI test instance is set up and operational.
* A BPI with at least one workstep defined.
* A workstep instance that is currently being executed by the BPI.

Test steps:

1. Get the identifier of the workstep instance that is currently being executed by the BPI.
2. Attempt to update the workstep instance while it is being executed by the BPI.
3. Verify that the workstep instance was not updated and that an appropriate error message is returned.

> 4. Wait for the workstep instance to complete execution.

Reviewer: this is covered by [R218]

Author: not quite R218 talks about a workstep, not a workstep instance. A workstep might be used in more than one workflow, and therefore, an instance of that workstep exists in more than one workflow. And when an instance is being executed it cannot be updated e.g. because the workstep was updated while a workstep instance is running.

5. Verify that the workstep instance can now be updated.

Test Passing criteria:

* Step 3 must return an appropriate error message indicating that the workstep instance cannot be updated while it is being executed by the BPI.
* Step 5 must successfully update the workstep instance.

#### **[R220]** 

> A workstep MUST be versioned within a BPI.

Reviewer: If the workstep has the same functionality after the change, does it need to have a different version?

E.g. changes to functional code that merely affect readability or code size, such as minification or changing the use of (non-significant) whitespace

Author: debatable ... changing to "functional change" to make it more explicit. It is an implementer decision where the versioning cut-off is.

*Note that versions of the same workstep do not have to be compatible with one another.* 

[[R220]](#r220) Testability:

Preconditions:

* A BPI test instance is set up and operational.
* A BPI has been created and contains at least one workstep.
* The BPI allows for versioning of worksteps.

Test Steps:

1. Create a new workstep within the BPI.

> 2. Verify that the workstep has been assigned a version number (e.g., 1.0).

Reviewer: what's wrong with alphanumeric identifiers? Nothing in the requirement specifies a number

3. Make a change to the workstep and save the changes.
4. Verify that the workstep's version number has been updated (e.g., 1.1).
5. Repeat steps 3-4 for at least two more changes to the workstep.
6. Verify that each change results in an updated version number.
7. Attempt to execute the workstep with an earlier version number (e.g., 1.0).
8. Verify that an error or warning is thrown indicating that the workstep version is outdated and needs to be updated.
9. Create a new version of the workstep.
10. Verify that the new version has a unique version number.
11. Attempt to execute the workstep with the new version number.
12. Verify that the workstep executes successfully.

Test Passing Criteria:

* The workstep versioning functionality allows for changes to be made to a workstep while maintaining a record of the changes and ensuring that an outdated version cannot be executed.
* The test passes if all steps execute successfully and the workstep can be versioned and executed without errors or warnings.

#### **[R221]** 
A workstep MUST be executed inside a BPI's Virtual State Machine (VSM).

> Note, that a BPI's Virtual State Machine is part of the BPI Processing Layer as discussed before.

Reviewer: before - where? Please link.

Author: done

[[CR24]<[O3]](#cr24o3) Testability:

Preconditions:

* A BPI test instance with the ability to create Output Pledges is set up and running.
* A workstep output has been used in a commercial value-creation event and an Output Pledge has been created for it.

Test Steps:

1. Verify that only the owners of the workstep output used in the commercial value-creation event are able to update the Output Pledge.
2. Attempt to update the Output Pledge using credentials of a non-owner of the workstep output.
3. Verify that the attempt to update the Output Pledge failed and the Output Pledge remained unchanged.
4. Attempt to update the Output Pledge using credentials of an owner of the workstep output.
5. Verify that the attempt to update the Output Pledge succeeded and the Output Pledge was updated accordingly.

Test Passing Criteria:

* The test is considered passed if the attempt to update the Output Pledge using credentials of a non-owner of the workstep output fails in Step 2, and the attempt to update the Output Pledge using credentials of an owner of the workstep output succeeds in Step 4, and the Output Pledge is updated accordingly in Step 5.

## 6.2 BPI Workflow

> After specifying a workstep, this document will now turn to a workflow.

Author: the document does not turn into a workflow ;-) @Ybittan

* The stored proofs, state objects, associated data, and their histories can be retrieved from the Storage capability.
* The retrieved information matches the original data and corresponds to the respective transactions.

#### **[R271]** 
The integrity of proofs, transactions, state objects, and their data and histories MUST be cryptographically verifiable by the owners of the accounts associated with the proofs, transactions, state objects, and their data and history.

[[R271]](#r271) Testability:

Preconditions:

* A BPI test system with the necessary components, including accounts, proofs, transactions, state objects, and their data and histories is operational.
* A cryptographic verification mechanisms with tooling to test integrity such as a Merkle Tree is properly configured and accessible to the account owners.
* One or more test transactions for the BPI.

Test Steps:

1. Generate a proof for a transaction on the BPI system.
2. Verify that the proof, along with the associated transaction, state object, and their data and histories, can be accessed by the respective account owners.

> 3. Using the chosen verfication mathod, cryptographically verify the integrity of the proof, transaction, state object, and their data and histories using the available tooling for the chosen verification method.

Reviewer 2: ```suggestion
3. Using the chosen verification method, cryptographically verify the integrity of the proof, transaction, state object, and their data and histories using the available tooling for the chosen verification method.
```

4. Confirm that the account owners are able to successfully verify the integrity of the aforementioned components.

Passing Criteria:

* A proof is generated for a transaction.
* The proof, along with the associated transaction, state object, and their data and histories, can be accessed by the respective account owners.
* Using the chosen verification method and associated tooling, the account owners can successfully cryptographically verify the integrity of the proof, transaction, state object, and their data and histories.

#### **[R272]** 
All updates to an agreement state and their associated accounts by a VSM MUST be communicated to all agreement counterparties through the Message capability in the BPI Middleware layer. 

[[R272]](#r272) Testability:

Preconditions:

* A BPI test system with the necessary components, including agreement states, associated accounts, and the Message capability in the BPI Middleware layer.
* All counterparties to a test agreement are registered as BPI Subjects in the test BPI and connected to the BPI Message capability.

> * There is one or more test transactions operating on the agreement state using a VSM.

Reviewer 2: ```suggestion
* There are one or more test transactions operating on the agreement state using a VSM.
```

Test Steps:

1. Submit one or more test transactions causing an update to an agreement state (state object) and its associated accounts using the VSM in the BPI system.
2. Verify that the generated state update message is processed by the Message capability.
3. Confirm that all agreement counterparties receive the state update message.
4. Validate that the state update message contains the relevant information about the update, such as the updated agreement state and the associated accounts.

Passing Criteria:

* A successful update is made to an agreement state and its associated accounts using the VSM.

> * The update is successfully by the Message capability.

Reviewer 2: ```suggestion
* The update is successfully executed by the Message capability.
```

*Note, see the BPI transaction lifecycle management flow in Fig. 10, in particular step (8).*

[[R273]](#r273) Testability:

Preconditions:

* An operational BPI test system with the necessary components, including the VSM, CCSM Abstraction Layer, CCSM, and the Message capability in the BPI Middleware layer.
* One or more test transactions that trigger state transition.

Test Steps:

1. Submit a test transaction to the BPI
2. Verify that a Proof-of-Correctness for a state transition and its associated data using the VSM in the BPI system was successfully generated.
3. Verify that the generated Proof-of-Correctness was sent to the CCSM Abstraction Layer through the Message capability in the BPI Middleware layer.
4. Verify that the CCSM Abstraction Layer receives the Proof-of-Correctness message.
5. Validate that the Proof-of-Correctness message contains the necessary information for subsequent commitment to the CCSM.
6. Confirm that the CCSM Abstraction Layer successfully commits the Proof-of-Correctness to the CCSM.
7. Repeat Step 1. through 6. for all test transactions.

Passing Criteria:

> * For each test transactions and associated state transition, a Proof-of-Correctness is successfully generated using the VSM.

Reviewer 2: ```suggestion
* For each test transaction and associated state transition, a Proof-of-Correctness is successfully generated using the VSM.
```

* The CCSM Abstraction Layer receives the Proof-of-Correctness message for each transaction.
* Each Proof-of-Correctness message contains the necessary information for subsequent commitment to the CCSM.
* The CCSM Abstraction Layer successfully commits each Proof-of-Correctness to the CCSM.

The following requirements are addressing the operating scenario where a BPI consists of more than one node. This is a perfectly feasible scenario with its pros and cons beyond the scope of this document to discuss. However, certain requirements need to be met for such a scenario to be operationally viable.

#### **[O7]** 
A BPI MAY consist of more than one processing node.

*This document will call such a structure a BPI network.*

[[O7]](#o7) Testability:

Preconditions:

* An operational BPI test system with multiple processing nodes.
* One or more test transactions.

Test Steps:

1. Verify that the BPI system has been configured with multiple processing nodes.
2. Submit a transaction to the BPI system.

> 3. Confirm that the transaction is distributed and processed across the multiple processing nodes.

Reviewer 2: ```suggestion
3. Confirm that the transaction is distributed and processed across multiple processing nodes.
```

Passing Criteria:

* The BPI network reaches consensus on the order of the transactions based on the submission order.
* The BPI network reaches consensus on the correct execution of the transactions.
* All nodes in the BPI network have the same ordered and executed transactions.

#### **[CR33]>[O7]** 
A BPI network MUST use a common execution framework. 

*Note that if more than one execution framework were chosen, no consensus could be reached on the outcome of a transaction because the state representation is execution framework dependent e.g. Ethereum account state vs. a zero-knowledge-proof of account state.*

[[CR33]>[O7]](#cr33o7) Testablity:

Preconditions:

* An operational BPI network with multiple nodes.
* One or more test transactions.

Test Steps:

1. Submit a set of transactions to the BPI network.
2. Monitor the execution of transactions across the network.
3. Validate that the transactions are executed on each node produce consistent and expected results.
3. Validate that the transactions that are executed on each node produce the same results on each node.

Passing Criteria:

* The transactions are executed on each node 

> * The transactions tha are executed on each node produce the same result.

Reviewer 2: ```suggestion
* The transactions that are executed on each node produce the same result.
```

In the case of a BPI, this document needs to distinguish between proof, transaction, and state data and its history and metadata associated with state objects, etc. Given the need to maintain consistency between state data on the CCSM utilized by the BPI and state data in the BPI, the following requirement is important.

#### **[R274]** 
Proof, transaction, and state object data and their histories together with their integrity proofs MUST be stored as partially persistent data in the BPI storage capability.

Partially persistent data is defined as append-only (write once, read many).

*The Storage capability requirements are defined in section [7 General BPI Storage Capabilities](#7-general-bpi-storage-capabilities).*

[[R274]](#r274) Testability:

Preconditions:

* An operational BPI test system.
* Data structures and formats for proof, transaction, and state object data, including their histories and integrity proofs.
* A test set for a proof, and a state object, including its history and (CCSM) integrity proof. 
* One or more test transactions and its expected results.

Test Steps:

1. Submit a  test transaction.

> 2. Verify that the transaction is correctly executed and the resulting proof, state object and history and integrity data is stored in the storage capability

Reviewer 2: ```suggestion
2. Verify that the transaction is correctly executed and the resulting proof, state object, and history and integrity data are stored in the storage capability
```

#### **[R315]** 
BPI Subjects participating in a workstep MUST agree upon the source and type of the external authoritative data used as input to a BPI workstep.

> [[R315]](#r15) Testability: An agreement on authoritative data sources will be embedded in a commercial document electronically represented on a BPI as required by [[R6]](#r6) to be bdining. And since [[R6]](#r6) is testable, [[R315]](#r15) is testable.

Reviewer 2: ```suggestion
[[R315]](#r15) Testability: An agreement on authoritative data sources will be embedded in a commercial document electronically represented on a BPI as required by [[R6]](#r6) to be bindining. And since [[R6]](#r6) is testable, [[R315]](#r15) is testable.
```
